pkg/controller/istio/helper.go

```diff
@@ -51,107 +55,130 @@ func CheckIstioEnabled(cfg *rest.Config) (bool, error) {
+	spec.Addresses = []string{host + "/32"}
+	spec.Location = istio.ServiceEntry_MESH_EXTERNAL
+	spec.Ports = []*istio.Port{
+		&istio.Port{
+			Name:     "TCP-" + portStr,
+			Number:   port,
+			Protocol: "TCP",
+		},
+	}
+	spec.Resolution = istio.ServiceEntry_NONE
 
-func buildVirtualServiceHTTP(name string, host string, port uint32) []byte {
-	portStr := strconv.Itoa(int(port))
+	serviceEntry := &istiov1alpha3.ServiceEntry{
+		Spec: spec,
+	}
+	serviceEntry.Name = name
+	serviceEntry.Namespace = os.Getenv(k8sutil.WatchNamespaceEnvVar)
 
-	return []byte(`{
-    "apiVersion": "networking.istio.io/v1alpha3",
-    "kind": "VirtualService",
-    "metadata": {
-        "name": "` + name + `",
-        "namespace": "` + os.Getenv(k8sutil.WatchNamespaceEnvVar) + `"
-    },
-    "spec": {
-        "hosts": [ "` + host + `" ],
-        "http": [{
-            "match": [{
-                "port": ` + portStr + `,
-                "headers": [{ "Host": "` + host + `" }]
-            }],
-            "route": [{
-                "destination": {
-                    "host": "` + host + `",
-                    "port": { "number": ` + portStr + ` }
-                }
-            }]
-        }]
-    }
-}`)
+	return serviceEntry
 }
 
-// BuildNameForEndpoint returns a name to be used as a base to identify Istio objects.
-func BuildNameForEndpoint(name string, host string, port uint32) string {
-	portStr := strconv.Itoa(int(port))
-	src := make([]byte, len(name)+len(host)+len(portStr))
-	src = strconv.AppendQuote(src, name)
-	src = strconv.AppendQuote(src, host)
-	src = strconv.AppendQuote(src, portStr)
+// BuildVirtualService returns an Istio VirtualService object for the given communication endpoint.
+func BuildVirtualService(name, host, protocol string, port uint32) *istiov1alpha3.VirtualService {
+	if net.ParseIP(host) != nil { // It's an IP.
+		return nil
+	}
 
-	sum := sha256.Sum256(src)
+	spec := istiov1alpha3.VirtualServiceSpec{}
+	spec.Hosts = []string{host}
 
+	switch protocol {
+	case "https":
+		spec.Tls = []*istio.TLSRoute{
+			&istio.TLSRoute{
+				Match: []*istio.TLSMatchAttributes{
+					&istio.TLSMatchAttributes{
+						SniHosts: []string{host},
+						Port:     port,
+					},
+				},
+				Route: []*istio.RouteDestination{
+					&istio.RouteDestination{
+						Destination: &istio.Destination{
+							Host: host,
+							Port: &istio.PortSelector{
+								Port: &istio.PortSelector_Number{Number: port},
+							},
+						},
+					},
+				},
+			},
+		}
+	case "http":
+		spec.Http = []*istio.HTTPRoute{
+			&istio.HTTPRoute{
+				Match: []*istio.HTTPMatchRequest{
+					&istio.HTTPMatchRequest{
+						Port: port,
+					},
+				},
+				Route: []*istio.HTTPRouteDestination{
+					&istio.HTTPRouteDestination{
+						Destination: &istio.Destination{
+							Host: host,
+							Port: &istio.PortSelector{
+								Port: &istio.PortSelector_Number{Number: port},
+							},
+						},
+					},
+				},
+			},
+		}
+	}
```

Comment:
Please move the generation of `spec.Tls` and `spec.Http` into functions to get this more readable.